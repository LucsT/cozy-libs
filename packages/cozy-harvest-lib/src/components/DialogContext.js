import { createContext, useContext } from 'react'

const DialogContext = createContext({})
export const useDialogContext = () => {
  return useContext(DialogContext)
}
export default DialogContext
