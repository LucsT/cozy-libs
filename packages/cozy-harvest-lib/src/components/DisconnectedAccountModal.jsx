import React, { useRef, useState } from 'react'
import { Contracts } from 'cozy-harvest-lib/dist/components/KonnectorConfiguration/ConfigurationTab/Contracts'
import KonnectorModalHeader from 'cozy-harvest-lib/dist/components/KonnectorModalHeader'

import { getCreatedByApp } from 'cozy-client/dist/models/utils'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'

import Modal, { ModalContent } from 'cozy-ui/transpiled/react/Modal'
import Card from 'cozy-ui/transpiled/react/Card'
import {
  Tabs,
  Tab,
  TabList,
  TabPanels,
  TabPanel
} from 'cozy-ui/transpiled/react/Tabs'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

import withLocales from './hoc/withLocales'
import { getAccountInstitutionLabel } from './KonnectorConfiguration/ConfigurationTab/bankAccountHelpers'

const createDummyKonnectorFromAccount = account => {
  return {
    name: getAccountInstitutionLabel(account),
    slug: getCreatedByApp(account) || null
  }
}

/**
 * Serves when configuring vendor accounts whose io.cozy.accounts
 * has been deleted.
 *
 * The UI must be as close as possible to the AccountModal but since
 * we lack a lot of data to show an AccountModal (no trigger, no konnector,
 * no account), it was implemented as a separate component.
 */
const DisconnectedModal = ({ accounts, onClose, initialActiveTab }) => {
  const { t } = useI18n()
  // We keep the konnector in a ref so that when we remove all accounts,
  // we still have a konnector to show the icon
  const konnectorRef = useRef()
  if (!konnectorRef.current) {
    konnectorRef.current = createDummyKonnectorFromAccount(accounts[0])
  }
  const { isMobile } = useBreakpoints()

  const [activeTab, setActiveTab] = useState(initialActiveTab)

  return (
    <Modal mobileFullscreen={true} dismissAction={onClose}>
      <KonnectorModalHeader konnector={konnectorRef.current} />
      <ModalContent className={isMobile ? 'u-p-0' : null}>
        <Tabs initialActiveTab={initialActiveTab}>
          <TabList>
            <Tab onClick={() => setActiveTab('data')} name="data">
              {t('modal.tabs.data')}
            </Tab>
            <Tab
              onClick={() => setActiveTab('configuration')}
              name="configuration"
            >
              {t('modal.tabs.configuration')}
            </Tab>
          </TabList>
        </Tabs>
      </ModalContent>
      <TabPanels activeTab={activeTab}>
        <TabPanel name="data" className="u-pt-1">
          <ModalContent>
            <Card>{t('disconnectedAccountModal.disconnected-help')}</Card>
          </ModalContent>
        </TabPanel>
        <TabPanel name="configuration">
          {accounts.length ? (
            <ModalContent className={isMobile ? 'u-p-0' : null}>
              <Contracts contracts={accounts} />
            </ModalContent>
          ) : (
            t('contracts.no-contracts')
          )}
        </TabPanel>
      </TabPanels>
    </Modal>
  )
}

DisconnectedModal.defaultProps = {
  initialActiveTab: 'configuration'
}

export default withLocales(DisconnectedModal)
