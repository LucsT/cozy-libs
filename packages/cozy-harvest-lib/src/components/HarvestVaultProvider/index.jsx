import React from 'react'
import PropTypes from 'prop-types'
import { withClient } from 'cozy-client'
import { VaultProvider } from 'cozy-keys-lib'

class HarvestVaultProvider extends React.PureComponent {
  render() {
    const { children, client } = this.props

    const vaultInstanceUrl = client.getStackClient().uri

    return <VaultProvider instance={vaultInstanceUrl}>{children}</VaultProvider>
  }
}

HarvestVaultProvider.propTypes = {
  client: PropTypes.object
}

export default withClient(HarvestVaultProvider)
