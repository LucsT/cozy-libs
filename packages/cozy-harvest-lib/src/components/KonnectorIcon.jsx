import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { withClient } from 'cozy-client'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'

// TODO move this to cozy-ui
class KonnectorIcon extends PureComponent {
  fetchIcon() {
    const { konnector, client } = this.props
    return client.getStackClient().getIconURL({
      type: 'konnector',
      slug: konnector.slug
    })
  }

  render() {
    // eslint-disable-next-line no-unused-vars
    const { client, konnector, ...restProps } = this.props
    return <AppIcon fetchIcon={this.fetchIcon.bind(this)} {...restProps} />
  }
}

KonnectorIcon.propTypes = {
  konnector: PropTypes.object.isRequired,
  client: PropTypes.object.isRequired
}
export default withClient(KonnectorIcon)
