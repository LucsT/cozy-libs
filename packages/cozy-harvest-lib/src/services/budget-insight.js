/**
 * Interface between Budget Insight and Cozy.
 *
 * - Deals with the konnector to get temporary tokens
 * - Converts low-level BI errors to KonnectorJobErrors
 */

import get from 'lodash/get'
import omit from 'lodash/omit'
import clone from 'lodash/clone'
import set from 'lodash/set'
import defaults from 'lodash/defaults'

import { waitForRealtimeEvent } from './jobUtils'
import { createBIConnection, updateBIConnection } from './bi-http'
import assert from '../assert'
import { mkConnAuth, biErrorMap } from 'cozy-bi-auth'
import biPublicKeyProd from './bi-public-key-prod.json'
import { KonnectorJobError } from '../helpers/konnectors'
import { LOGIN_SUCCESS_EVENT } from '../models/ConnectionFlow'
import logger from '../logger'

const DECOUPLED_ERROR = 'decoupled'
const ADDITIONAL_INFORMATION_NEEDED_ERROR = 'additionalInformationNeeded'

const configsByMode = {
  prod: {
    mode: 'prod',
    url: 'https://cozy.biapi.pro/2.0',
    publicKey: biPublicKeyProd
  },
  dev: {
    mode: 'dev',
    url: 'https://cozytest-sandbox.biapi.pro/2.0'
  }
}

const getBIConnectionIdFromAccount = account =>
  get(account, 'data.auth.bi.connId')

/**
 * Converts and chains error
 */
const convertBIErrortoKonnectorJobError = error => {
  const errorCode = error ? error.code : null
  const cozyErrorMessage = errorCode ? biErrorMap[errorCode] : null
  const errorMessage =
    cozyErrorMessage ||
    (errorCode ? `UNKNOWN_ERROR.${errorCode}` : 'UNKNOWN_ERROR')
  const err = new KonnectorJobError(errorMessage)
  err.original = error
  throw err
}

const getBIModeFromCozyURL = rawCozyURL => {
  if (!rawCozyURL) {
    return 'dev'
  }
  const cozyURL = new URL(rawCozyURL)
  const domain = cozyURL.host
    .split('.')
    .slice(-2)
    .join('.')
    .split(':')
    .shift()
  switch (domain) {
    case 'cozy.tools':
    case 'cozy.wtf':
    case 'cozy.blue':
    case 'cozy.works':
    case 'cozy.red':
    case 'cozy.company':
    case 'cozy.solutions':
    case 'toutatice.fr':
    case 'cozymaif.cloud':
    case 'cozy-maif-int.fr':
    case 'cozy.dev':
      return 'dev'
    default:
      return 'prod'
  }
}

export const getBIConfigForCozyURL = url => {
  const mode = getBIModeFromCozyURL(url)
  if (configsByMode[mode]) {
    return configsByMode[mode]
  } else {
    throw new Error(
      `getBIConfig: Unknown mode ${mode}, known modes are ${Object.keys(
        configsByMode
      ).join(', ')}`
    )
  }
}

export const isBudgetInsightConnector = konnector => {
  return (
    konnector.partnership &&
    konnector.partnership.domain.includes('budget-insight')
  )
}

const createTemporaryToken = async ({ client, account, konnector }) => {
  assert(
    account && account._id,
    'createTemporaryToken: Account without id passed to createTemporaryToken'
  )
  assert(konnector.slug, 'createTemporaryToken: No konnector slug')
  const jobResponse = await client.stackClient.jobs.create('konnector', {
    mode: 'getTemporaryToken',
    konnector: konnector.slug,
    account: account._id,
    bankId: account.auth.bankId
  })
  const event = await waitForRealtimeEvent(
    client,
    jobResponse.data.attributes,
    'result'
  )
  return event.data.result.code
}

export const saveTemporaryToken = (flow, tempToken) =>
  flow.setData({
    tempToken
  })

export const getTemporaryToken = flow => {
  const data = flow.getData()
  return data.tempToken
}

/**
 * Ensures a BI connection is ready
 *
 *  - Calls the getBITemporaryToken mode of the konnector
 *
 * Then:
 * If no BI connection is present in the io.cozy.accounts:
 *  - Creates the BI connection using user supplied data
 * If a BI connection exists:
 *  - Updates the BI connection using user supplied data
 *
 * @param  {Client} options.client
 * @param  {io.cozy.accounts} options.account
 * @param  {io.cozy.konnectors} options.konnector
 * @return {BIConnection}
 */
export const createOrUpdateBIConnection = async ({
  account,
  client,
  konnector,
  flow
}) => {
  const config = getBIConfigForCozyURL(client.stackClient.uri)
  const connId = getBIConnectionIdFromAccount(account)
  logger.info('Creating temporary token...')

  const tempToken = await createTemporaryToken({
    account,
    client,
    konnector
  })
  saveTemporaryToken(flow, tempToken)

  logger.info('Created temporary token')
  assert(tempToken, 'No temporary token')
  const credentials = { ...account.auth }
  // The konnector can have "baked-in" parameters that need to be passed in the
  // auth. This is the case for example for some konnectors for the bankId
  // parameter
  defaults(credentials, konnector.parameters)
  const credsToSend = await mkConnAuth(config, credentials)

  try {
    logger.info('Creating connection...')
    const connection = await (connId
      ? updateBIConnection(config, connId, credsToSend, tempToken)
      : createBIConnection(config, credsToSend, tempToken))

    logger.info(`Created connection ${connection.id}`)
    return connection
  } catch (e) {
    return convertBIErrortoKonnectorJobError(e)
  }
}

const SENSIBLE_FIELDS = ['password', 'secret', 'code', 'dob']
const removeSensibleDataFromAccount = fullAccount => {
  const account = clone(fullAccount)
  account.auth = omit(account.auth, SENSIBLE_FIELDS)
  return account
}

export const setBIConnectionId = (originalAccount, biConnectionId) => {
  const account = clone(originalAccount)
  set(account, 'data.auth.bi.connId', biConnectionId)
  return account
}

export const getBIConnectionId = account => {
  return get(account, 'data.auth.bi.connId')
}

export const updateBIConnectionFromFlow = async (flow, connectionData) => {
  const client = flow.client
  const account = flow.account

  const config = getBIConfigForCozyURL(client.stackClient.uri)
  const connId = getBIConnectionIdFromAccount(account)
  const temporaryToken = getTemporaryToken(flow)

  logger.debug('Updating BI connection')
  const updatedConnection = await updateBIConnection(
    config,
    connId,
    connectionData,
    temporaryToken
  )
  logger.info('Updated BI connection')

  flow.setData({ biConnection: updatedConnection })
}

export const sendAdditionalInformation = (flow, fields) => {
  return updateBIConnectionFromFlow(flow, fields)
}

export const resumeBIConnection = flow => {
  return updateBIConnectionFromFlow(flow, { resume: 'true' })
}

/**
 * Checks for 2FA or decoupled requests
 *
 * - Resolves when 2FA has been resolved
 * - Triggers LOGIN_SUCCESS_EVENT
 */
export const finishConnection = async ({ biConnection, flow }) => {
  if (biConnection.error) {
    if (biConnection.error === DECOUPLED_ERROR) {
      const twoFAOptions = { type: 'app', retry: false }
      await flow.saveTwoFARequest(twoFAOptions)
      logger.debug('Resuming BI connection...')
      await resumeBIConnection(flow)
      flow.triggerEvent(LOGIN_SUCCESS_EVENT)
      logger.debug('Finished waiting for decoupled connection to be validated')
    } else if (biConnection.error === ADDITIONAL_INFORMATION_NEEDED_ERROR) {
      const twoFAOptions = { type: 'sms', retry: false }
      while (
        flow.getData().biConnection.error === 'additionalInformationNeeded'
      ) {
        await flow.saveTwoFARequest(twoFAOptions)
        await flow.waitForTwoFA()
      }
      flow.triggerEvent(LOGIN_SUCCESS_EVENT)
    } else {
      throw convertBIErrortoKonnectorJobError(biConnection.error)
    }
  } else {
    flow.triggerEvent(LOGIN_SUCCESS_EVENT)
  }
}

/**
 * Used in the custom konnector policy for BI
 */
export const onBIAccountCreation = async ({
  account,
  client,
  flow,
  konnector,
  createOrUpdateBIConnectionFn = createOrUpdateBIConnection
}) => {
  const fullAccount = account
  account = removeSensibleDataFromAccount(account)

  account = await flow.saveAccount(account)

  const biConnection = await createOrUpdateBIConnectionFn({
    account: {
      ...fullAccount,
      _id: account._id
    },
    client,
    konnector,
    flow
  })

  flow.setData({
    biConnection
  })

  account = await flow.saveAccount(setBIConnectionId(account, biConnection.id))

  // At this point, we can be in two fa request stat
  await finishConnection({
    account,
    biConnection,
    client,
    flow
  })

  return account
}

export const getAdditionalInformationNeeded = flow => {
  const biConnection = flow.getData().biConnection
  assert(
    biConnection,
    'No BI connection saved in connection flow, cannot determine additional fields'
  )
  return biConnection.fields
}

export const konnectorPolicy = {
  name: 'budget-insight',
  match: isBudgetInsightConnector,
  saveInVault: false,
  onAccountCreation: onBIAccountCreation,
  sendAdditionalInformation: sendAdditionalInformation,
  getAdditionalInformationNeeded
}
