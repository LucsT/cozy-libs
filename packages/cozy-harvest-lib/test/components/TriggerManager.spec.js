/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { render } from '@testing-library/react'

import { DumbTriggerManager as TriggerManager } from 'components/TriggerManager'
import cronHelpers from 'helpers/cron'
import { konnectorPolicy as biKonnectorPolicy } from '../../src/services/budget-insight'

jest.mock('cozy-flags', () => name => {
  if (name === 'bi-konnector-policy') {
    return true
  } else {
    return false
  }
})

jest.mock('cozy-keys-lib')

jest.mock('cozy-doctypes', () => {
  const doctypes = jest.requireActual('cozy-doctypes')

  const CozyFolder = {
    copyWithClient: () => CozyFolder,
    ensureMagicFolder: () => ({ path: '/Administrative' }),
    magicFolders: {
      ADMINISTRATIVE: 'io.cozy.apps/administrative',
      PHOTOS: '/photos'
    }
  }

  return {
    ...doctypes,
    CozyFolder
  }
})

jest.mock('../../src/services/budget-insight', () => {
  const originalBudgetInsight = jest.requireActual(
    '../../src/services/budget-insight'
  )
  return {
    konnectorPolicy: {
      ...originalBudgetInsight.konnectorPolicy,
      onAccountCreation: jest.fn()
    }
  }
})

jest.mock('cozy-ui/transpiled/react/utils/color')

const fixtures = {
  data: {
    username: 'foo',
    passphrase: 'bar'
  },
  konnector: {
    slug: 'konnectest',
    fields: {
      username: {
        type: 'text'
      },
      passphrase: {
        type: 'password'
      }
    }
  },
  konnectorWithFolder: {
    name: 'myBills',
    slug: 'mybills',
    fields: {
      advancedFields: {
        folderPath: {
          advanced: true
        }
      }
    }
  },
  folder: {
    _id: '3f5b288af36041f189ec22063adab706'
  },
  folderPath: '/Administrative/myBills/foo',
  folderPermission: {
    saveFolder: {
      type: 'io.cozy.files',
      values: ['3f5b288af36041f189ec22063adab706'],
      verbs: ['GET', 'PATCH', 'POST']
    }
  },
  triggerAttributes: {
    arguments: '0 0 0 * * 0',
    type: '@cron',
    worker: 'konnector',
    message: {
      account: 'created-account-id',
      konnector: 'konnectest'
    }
  },
  account: {
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'bar'
    },
    identifier: 'username'
  },
  createdAccount: {
    _id: 'created-account-id',
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'bar'
    },
    identifier: 'username'
  },
  updatedAccount: {
    _id: 'updated-account-id',
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'fuz'
    }
  },
  existingAccount: {
    _id: 'existing-account-id',
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'bar'
    },
    identifier: 'username'
  },
  existingTrigger: {
    id: 'existing-trigger-id',
    _type: 'io.cozy.triggers',
    attributes: {
      arguments: '0 0 0 * * 0',
      type: '@cron',
      worker: 'konnector',
      message: {
        account: 'existing-account-id',
        konnector: 'konnectest'
      }
    }
  },
  bankingKonnectorAccountAttributes: {
    auth: {
      identifier: '80564789',
      secret: '13337'
    },
    identifier: '80564789'
  },
  createdTrigger: {
    id: 'created-trigger-id',
    _type: 'io.cozy.triggers',
    attributes: {
      arguments: '0 0 0 * * 0',
      type: '@cron',
      worker: 'konnector',
      message: {
        account: 'updated-account-id',
        konnector: 'konnectest'
      }
    }
  },
  launchedJob: {
    type: 'io.cozy.jobs',
    id: 'lauched-job-id',
    domain: 'cozy.tools:8080',
    worker: 'konnector',
    state: 'running',
    queued_at: '2016-09-19T12:35:08Z',
    started_at: '2016-09-19T12:35:08Z',
    error: '',
    links: {
      self: '/jobs/lauched-job-id'
    }
  },
  runningJob: {
    type: 'io.cozy.jobs',
    id: 'running-job-id',
    domain: 'cozy.tools:8080',
    worker: 'konnector',
    state: 'running',
    queued_at: '2016-09-19T12:35:08Z',
    started_at: '2016-09-19T12:35:08Z',
    error: '',
    links: {
      self: '/jobs/running-job-id'
    }
  },
  doneJob: {
    type: 'io.cozy.jobs',
    id: 'done-job-id',
    domain: 'cozy.tools:8080',
    worker: 'konnector',
    state: 'done',
    queued_at: '2016-09-19T12:35:08Z',
    started_at: '2016-09-19T12:35:08Z',
    error: '',
    links: {
      self: '/jobs/done-job-id'
    }
  }
}

const mockVaultClient = {
  createNewCipher: jest.fn(),
  saveCipher: jest.fn(),
  getByIdOrSearch: jest.fn(),
  decrypt: jest.fn(),
  createNewCozySharedCipher: jest.fn(),
  get: jest.fn(),
  getAll: jest.fn(),
  getAllDecrypted: jest.fn(),
  shareWithCozy: jest.fn(),
  isLocked: jest.fn().mockResolvedValue(false)
}

const addPermissionMock = jest.fn()
const addReferencesToMock = jest.fn()
const createTriggerMock = jest.fn()
const createDirectoryByPathMock = jest.fn()
const statDirectoryByPathMock = jest.fn()
const launchTriggerMock = jest.fn()
const saveAccountMock = jest.fn()

const tMock = jest.fn()

const props = {
  addPermission: addPermissionMock,
  addReferencesTo: addReferencesToMock,
  konnector: fixtures.konnector,
  createTrigger: createTriggerMock,
  createDirectoryByPath: createDirectoryByPathMock,
  statDirectoryByPath: statDirectoryByPathMock,
  saveAccount: saveAccountMock,
  launch: launchTriggerMock,
  t: tMock,
  vaultClient: mockVaultClient,
  breakpoints: { isMobile: false },
  onVaultDismiss: jest.fn()
}

const propsWithAccount = {
  ...props,
  account: fixtures.existingAccount,
  trigger: fixtures.existingTrigger
}

const defaultOnError = error => {
  throw error
}

const setup = ({
  konnector = fixtures.konnector,
  account,
  trigger,
  onError = defaultOnError
} = {}) => {
  const root = shallow(
    <TriggerManager
      {...props}
      konnector={konnector}
      trigger={trigger}
      account={account}
      onError={onError}
    />
  )
  return { root }
}

const shallowWithoutAccount = konnector => {
  const { root } = setup({ konnector })
  return root
}

const shallowWithAccount = options => {
  const { root } = setup({
    account: fixtures.existingAccount,
    trigger: fixtures.existingTrigger,
    ...options
  })
  return root
}

const isHigherComponentOfDisplayName = displayName => node => {
  const type = node.type()
  return type.displayName && type.displayName.includes(`(${displayName})`)
}

describe('TriggerManager', () => {
  beforeEach(() => {
    createTriggerMock.mockResolvedValue(fixtures.createdTrigger)
    saveAccountMock.mockResolvedValue(fixtures.createdAccount)
    mockVaultClient.createNewCozySharedCipher.mockResolvedValue({
      id: 'cipher-id-1'
    })
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  describe('when given an oauth konnector', () => {
    it('should redirect to OAuthForm', () => {
      const konnector = {
        oauth: {
          scope: 'test'
        }
      }
      const component = shallow(
        <TriggerManager {...props} konnector={konnector} />
      ).getElement()
      expect(component).toMatchSnapshot()
    })
  })

  describe('when given no account', () => {
    it('should render correctly', () => {
      const component = shallowWithoutAccount().getElement()
      expect(component).toMatchSnapshot()
    })

    describe('when the vault does not contain ciphers', () => {
      it('should show the new account form', async () => {
        mockVaultClient.getAll.mockResolvedValue([])
        mockVaultClient.getAllDecrypted.mockResolvedValue([])

        const { findByLabelText } = render(<TriggerManager {...props} />)

        const usernameField = await findByLabelText('username')
        const passwordField = await findByLabelText('passphrase')

        expect(usernameField).toBeDefined()
        expect(passwordField).toBeDefined()
      })
    })

    describe('when the vault contains ciphers', () => {
      it('should show the ciphers list', async () => {
        mockVaultClient.getAll.mockResolvedValue([{ id: 'cipher1' }])
        mockVaultClient.getAllDecrypted.mockResolvedValue([
          {
            id: 'cipher1',
            name: fixtures.konnector.name,
            login: {
              username: 'Isabelle'
            }
          }
        ])

        const { findByText } = render(<TriggerManager {...props} />)

        const cipherItem = await findByText('Isabelle')

        expect(cipherItem).toBeDefined()
      })
    })
  })

  describe('when given an account', () => {
    it('should render correctly', () => {
      const component = shallowWithAccount().getElement()
      expect(component).toMatchSnapshot()
    })

    describe('when the vault contains ciphers', () => {
      it('should show the account form', async () => {
        mockVaultClient.getAll.mockResolvedValue([])
        mockVaultClient.getAllDecrypted.mockResolvedValue([])

        const { findByLabelText } = render(
          <TriggerManager {...propsWithAccount} />
        )

        const usernameField = await findByLabelText('username')
        const passwordField = await findByLabelText('passphrase')

        expect(usernameField).toBeDefined()
        expect(passwordField).toBeDefined()
      })
    })

    describe('when the vault does not contain ciphers', () => {
      it('should show the account form', async () => {
        mockVaultClient.getAll.mockResolvedValue([{ id: 'cipher1' }])
        mockVaultClient.getAllDecrypted.mockResolvedValue([
          {
            id: 'cipher1',
            name: fixtures.konnector.name,
            login: {
              username: 'Isabelle'
            }
          }
        ])

        const { findByLabelText } = render(
          <TriggerManager {...propsWithAccount} />
        )

        const usernameField = await findByLabelText('username')
        const passwordField = await findByLabelText('passphrase')

        expect(usernameField).toBeDefined()
        expect(passwordField).toBeDefined()
      })
    })
  })

  describe('handleError', () => {
    beforeEach(() => {
      statDirectoryByPathMock.mockResolvedValue(fixtures.folder)
      createDirectoryByPathMock.mockResolvedValue(fixtures.folder)
    })

    it('should render error', async () => {
      const wrapper = shallowWithAccount({ onError: null })
      await wrapper.instance().handleError(new Error('Test error'))
      expect(wrapper.getElement()).toMatchSnapshot()
    })

    const clientMutations = {
      saveAccount: saveAccountMock,
      createDirectoryByPath: createDirectoryByPathMock,
      addPermission: addPermissionMock,
      addReferencesTo: addReferencesToMock,
      createTrigger: createTriggerMock,
      statDirectoryByPath: statDirectoryByPathMock
    }

    for (var mutation of Object.keys(clientMutations)) {
      const mutationSync = mutation
      it(`should be called when ${mutation} fails`, async () => {
        clientMutations[mutationSync].mockRejectedValue(
          new Error(`${mutationSync} error`)
        )

        if (mutationSync !== 'statDirectoryByPath') {
          statDirectoryByPathMock.mockResolvedValue(null)
        }

        const wrapper = shallowWithoutAccount(fixtures.konnectorWithFolder)

        jest
          .spyOn(wrapper.instance(), 'handleError')
          .mockImplementation(() => {})

        await wrapper.instance().handleSubmit(fixtures.data)
        expect(wrapper.instance().handleError).toHaveBeenCalledWith(
          new Error(`${mutationSync} error`)
        )
      })
    }
  })
})
