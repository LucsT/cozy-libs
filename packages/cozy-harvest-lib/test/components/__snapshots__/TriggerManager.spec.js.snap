// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerManager when given an oauth konnector should redirect to OAuthForm 1`] = `
<withI18n(withConnectionFlow(OAuthForm))
  flow={
    ConnectionFlow {
      "client": CozyClient {
        "appMetadata": Object {},
        "chain": CozyLink {
          "request": [Function],
        },
        "client": CozyStackClient {
          "jobs": JobCollection {
            "stackClient": CozyStackClient {
              "jobs": [Circular],
              "konnectors": KonnectorCollection {
                "doctype": "io.cozy.konnectors",
                "endpoint": "/konnectors/",
                "indexes": Object {},
                "stackClient": [Circular],
              },
              "options": Object {
                "onRevocationChange": [Function],
                "onTokenRefresh": [Function],
              },
              "token": null,
              "uri": "",
            },
          },
          "konnectors": KonnectorCollection {
            "doctype": "io.cozy.konnectors",
            "endpoint": "/konnectors/",
            "indexes": Object {},
            "stackClient": CozyStackClient {
              "jobs": JobCollection {
                "stackClient": [Circular],
              },
              "konnectors": [Circular],
              "options": Object {
                "onRevocationChange": [Function],
                "onTokenRefresh": [Function],
              },
              "token": null,
              "uri": "",
            },
          },
          "options": Object {
            "onRevocationChange": [Function],
            "onTokenRefresh": [Function],
          },
          "token": null,
          "uri": "",
        },
        "handleRevocationChange": [Function],
        "handleTokenRefresh": [Function],
        "idCounter": 1,
        "instanceOptions": Object {},
        "isLogged": false,
        "links": Array [
          StackLink {
            "stackClient": CozyStackClient {
              "jobs": JobCollection {
                "stackClient": [Circular],
              },
              "konnectors": KonnectorCollection {
                "doctype": "io.cozy.konnectors",
                "endpoint": "/konnectors/",
                "indexes": Object {},
                "stackClient": [Circular],
              },
              "options": Object {
                "onRevocationChange": [Function],
                "onTokenRefresh": [Function],
              },
              "token": null,
              "uri": "",
            },
          },
        ],
        "options": Object {},
        "plugins": Object {},
        "schema": Schema {
          "byDoctype": Object {},
          "client": CozyStackClient {
            "jobs": JobCollection {
              "stackClient": [Circular],
            },
            "konnectors": KonnectorCollection {
              "doctype": "io.cozy.konnectors",
              "endpoint": "/konnectors/",
              "indexes": Object {},
              "stackClient": [Circular],
            },
            "options": Object {
              "onRevocationChange": [Function],
              "onTokenRefresh": [Function],
            },
            "token": null,
            "uri": "",
          },
        },
        "stackClient": CozyStackClient {
          "jobs": JobCollection {
            "stackClient": [Circular],
          },
          "konnectors": KonnectorCollection {
            "doctype": "io.cozy.konnectors",
            "endpoint": "/konnectors/",
            "indexes": Object {},
            "stackClient": [Circular],
          },
          "options": Object {
            "onRevocationChange": [Function],
            "onTokenRefresh": [Function],
          },
          "token": null,
          "uri": "",
        },
      },
      "data": Object {},
      "getKonnectorSlug": [Function],
      "getTwoFACodeProvider": [Function],
      "handleAccountUpdated": [Function],
      "handleJobUpdated": [Function],
      "handleTwoFA": [Function],
      "konnector": Object {
        "oauth": Object {
          "scope": "test",
        },
      },
      "launch": [Function],
      "realtime": CozyRealtime {
        "client": CozyClient {
          "appMetadata": Object {},
          "chain": CozyLink {
            "request": [Function],
          },
          "client": CozyStackClient {
            "jobs": JobCollection {
              "stackClient": CozyStackClient {
                "jobs": [Circular],
                "konnectors": KonnectorCollection {
                  "doctype": "io.cozy.konnectors",
                  "endpoint": "/konnectors/",
                  "indexes": Object {},
                  "stackClient": [Circular],
                },
                "options": Object {
                  "onRevocationChange": [Function],
                  "onTokenRefresh": [Function],
                },
                "token": null,
                "uri": "",
              },
            },
            "konnectors": KonnectorCollection {
              "doctype": "io.cozy.konnectors",
              "endpoint": "/konnectors/",
              "indexes": Object {},
              "stackClient": CozyStackClient {
                "jobs": JobCollection {
                  "stackClient": [Circular],
                },
                "konnectors": [Circular],
                "options": Object {
                  "onRevocationChange": [Function],
                  "onTokenRefresh": [Function],
                },
                "token": null,
                "uri": "",
              },
            },
            "options": Object {
              "onRevocationChange": [Function],
              "onTokenRefresh": [Function],
            },
            "token": null,
            "uri": "",
          },
          "handleRevocationChange": [Function],
          "handleTokenRefresh": [Function],
          "idCounter": 1,
          "instanceOptions": Object {},
          "isLogged": false,
          "links": Array [
            StackLink {
              "stackClient": CozyStackClient {
                "jobs": JobCollection {
                  "stackClient": [Circular],
                },
                "konnectors": KonnectorCollection {
                  "doctype": "io.cozy.konnectors",
                  "endpoint": "/konnectors/",
                  "indexes": Object {},
                  "stackClient": [Circular],
                },
                "options": Object {
                  "onRevocationChange": [Function],
                  "onTokenRefresh": [Function],
                },
                "token": null,
                "uri": "",
              },
            },
          ],
          "options": Object {},
          "plugins": Object {},
          "schema": Schema {
            "byDoctype": Object {},
            "client": CozyStackClient {
              "jobs": JobCollection {
                "stackClient": [Circular],
              },
              "konnectors": KonnectorCollection {
                "doctype": "io.cozy.konnectors",
                "endpoint": "/konnectors/",
                "indexes": Object {},
                "stackClient": [Circular],
              },
              "options": Object {
                "onRevocationChange": [Function],
                "onTokenRefresh": [Function],
              },
              "token": null,
              "uri": "",
            },
          },
          "stackClient": CozyStackClient {
            "jobs": JobCollection {
              "stackClient": [Circular],
            },
            "konnectors": KonnectorCollection {
              "doctype": "io.cozy.konnectors",
              "endpoint": "/konnectors/",
              "indexes": Object {},
              "stackClient": [Circular],
            },
            "options": Object {
              "onRevocationChange": [Function],
              "onTokenRefresh": [Function],
            },
            "token": null,
            "uri": "",
          },
        },
        "onClientLogin": [Function],
        "onClientLogout": [Function],
        "onClientTokenRefreshed": [Function],
        "onDeviceReady": [Function],
        "onOffline": [Function],
        "onOnline": [Function],
        "onResume": [Function],
        "onVisibilityChange": [Function],
        "onWebSocketClose": [Function],
        "onWebSocketError": [Function],
        "onWebSocketMessage": [Function],
        "onWebSocketOpen": [Function],
        "retryManager": RetryManager {
          "_events": Object {
            "error": Array [
              [Function],
            ],
          },
          "baseWaitAfterFirstFailure": 128,
          "maxWaitBetweenRetries": 300000,
          "onFailure": [Function],
          "onSuccess": [Function],
          "onSuccessAcknowledged": [Function],
          "raiseErrorAfterAttempts": 8,
          "retries": 0,
          "timeBeforeSuccessful": 1200,
          "wait": 0,
        },
        "subscriptions": SubscriptionList {
          "subscriptions": Array [],
        },
      },
      "sendTwoFACode": [Function],
      "state": Object {
        "status": "IDLE",
      },
      "trigger": undefined,
      "twoFAWaiters": Array [],
      "unsubscribeAllRealtime": null,
      "unwatch": [Function],
    }
  }
  konnector={
    Object {
      "oauth": Object {
        "scope": "test",
      },
    }
  }
  onSuccess={[Function]}
/>
`;
